### PR TITLE
Scipy 1.0 Roadmap

### PR DESCRIPTION
**Pull request by rgommers**
Note: to be discussed on the list, don't merge.

### CODE REVIEW

File: doc/ROADMAP.rst.txt
Comment (sciunto) [inline doc/ROADMAP.rst.txt]: That's not clear to me. Should we
* open a PR and then discuss on the ML
* open a PR, announce on the ML, and discuss on the PR
* discuss on the ML and once a consensus is established, open a PR

There's probably other combinations :)

Comment (argriffing) [inline doc/ROADMAP.rst.txt]: I'd say open a PR, then if it's deemed to be too disruptive someone will suggest to discuss on the mailing list before merging.  It might also be suggested to discuss on the mailing list if you have not yet begun work on the PR, with the idea being to get feedback about whether the hypothetical future PR would even be considered, but in practice this may not be useful if you are just told that "there is really no danger to *considering* a proposal".

File: doc/ROADMAP.rst.txt
Comment (ev-br) [inline doc/ROADMAP.rst.txt]: Been a while :-). I played a bit with wrapping `std::map`, and it seems reasonable so far. Comments welcome: https://github.com/ev-br/sparr
(this is outside of scipy for now just for convenience, can be moved to the scipy umberlla if desired)

**Comment by rgommers** [comment]
Okay, I want to merge this PR, as mentioned on the mailing list. It will never be perfect before we hit the green button, so I will update it, then ping here again, and aim to merge it by 19 Sep (2 weeks from now). Then we can improve it further after it's merged.

Alternatively, we close it if it's deemed not useful (enough). 3 years is long enough for this to be open.

File: doc/ROADMAP.rst.txt
Comment (endolith) [inline doc/ROADMAP.rst.txt]: @rifasaurousrex I was going to propose that the existing `signal.convolve()` have a `mode='circ'` option, but I wasn't sure if there was a practical use for it.

File: doc/ROADMAP.rst.txt
Comment (endolith) [inline doc/ROADMAP.rst.txt]: split into cwt and dwt?

File: CONTRIBUTING
Comment (ev-br) [inline CONTRIBUTING]: Might want to drop the 1.0 qualifier. The document seems to evolve into a more general roadmap for the scipy project.

File: doc/ROADMAP.rst.txt
Comment (ev-br) [inline doc/ROADMAP.rst.txt]: There is some more work by @nmayorov, maybe there's something to add, esp as new features?

File: doc/ROADMAP.rst.txt
Comment (ev-br) [inline doc/ROADMAP.rst.txt]: Might mention `approx_derivative` which is for now private in scipy.optimize. I think it could serve as a basis for scipy.diff rather than (or as an alternative to) numdifftools.

File: doc/ROADMAP.rst.txt
Comment (Eric89GXL) [inline doc/ROADMAP.rst.txt]: > I'd say open a PR, then if it's deemed to be too disruptive someone will suggest to discuss on the mailing list before merging

Sounds good to me. But we might want to say something along the lines that for larger PRs people might want to start a discussion first in case the proposal is substantially altered (or rejected) so that they don't waste coding time.

File: doc/ROADMAP.rst.txt
Comment (Eric89GXL) [inline doc/ROADMAP.rst.txt]: It should be straightforward enough to add to the `convolve` docs (somewhere!) an example of how to do circular convolution as something like `ifft(fft(x) * fft(h))`. I agree it doesn't need to be on the 1.0 roadmap, though. @rifasaurousrex can you open a separate issue about it and we can decide how to deal with it?

File: doc/ROADMAP.rst.txt
Comment (Eric89GXL) [inline doc/ROADMAP.rst.txt]: Do we need more than what #5608 provided? If so this should be updated to make it a bit clearer (cross-refs, more docs...?)

File: doc/ROADMAP.rst.txt
Comment (Eric89GXL) [inline doc/ROADMAP.rst.txt]: I agree the output of `rfft` should at least optionally be able to be complex

**Comment by Eric89GXL** [comment]
+1 for merge once the last few comments are addressed. I'll try to keep these in mind when people work on PRs in case bits of this can be whittled away.

File: doc/ROADMAP.rst.txt
Comment (person142) [inline doc/ROADMAP.rst.txt]: Though there are still a lot of functions that need improvements in precision, probably the only show-stoppers are hypergeometric functions, parabolic cylinder functions, and spheroidal wave functions. Three possible ways to handle this:

 - Get good double-precision implementations. This is doable for parabolic cylinder functions (in progress). I think it's possible for hypergeometric functions, though maybe not in time. For spheroidal wavefunctions this is not possible with current theory.

- Port Boost's arbitrary precision library and use it under the hood to get double precision accuracy. This might be necessary as a stopgap measure for hypergeometric functions; the idea of using arbitrary precision has been suggested before by @nmayorov and in https://github.com/scipy/scipy/issues/5349#issuecomment-244542364. I think this is necessary for spheroidal wave functions, but thankfully someone has already done it: https://github.com/radelman/scattering.

- Add clear warnings to the documentation about the limits of the existing implementations.

File: doc/ROADMAP.rst.txt
Comment (person142) [inline doc/ROADMAP.rst.txt]: We might also want to establish an official policy for handling gradual loss of precision. Currently some routines return best guess (most hypergeometric functions, `digamma`, and others) and some return `nan` (Bessel functions, Struve functions, and others). I like returning best guess and warning: often inaccurate regions are near zeros, and a small less precise result might not pollute a larger computation, but a `nan` will.

File: doc/ROADMAP.rst.txt
Comment (aeberspaecher) [inline doc/ROADMAP.rst.txt]: To me it seems chirp z-transforms are in SciPy's scope. @endolith has started work in PR #4607. I would love to see this merged in the future (and possibly extended to 2d/3d/Nd).

File: doc/ROADMAP.rst.txt
Comment (shoyer) [inline doc/ROADMAP.rst.txt]: Removing the Python KDTree and renaming cKDTree -> KDTree (with an alias for backwards compatibility) could be appropriate for SciPy 1.0. Names like "cKDTree" are very outdated these names.

File: doc/ROADMAP.rst.txt
Comment (rgommers) [inline doc/ROADMAP.rst.txt]: I'm actually considering removing this. PyWavelets is now in pretty good shape, so could be used for this. However, it seems to be doing well as a standalone project and is not difficult to install. So maybe we should just leave things the way they are, and consider pulling it into SciPy only if development stalls for whatever reason.

File: CONTRIBUTING
Comment (rgommers) [inline CONTRIBUTING]: Good point, will drop it.

File: doc/ROADMAP.rst.txt
Comment (pv) [inline doc/ROADMAP.rst.txt]: Pulling it into scipy makes sense only if the maintainers of that 
project want to do it.
.
If the wavelet functionality is not fixed up, it may make sense to 
deprecate the stuff currently in scipy.

File: doc/ROADMAP.rst.txt
Comment (rgommers) [inline doc/ROADMAP.rst.txt]: Thanks @person142, that's excellent. Will put that in verbatim.

File: doc/ROADMAP.rst.txt
Comment (rgommers) [inline doc/ROADMAP.rst.txt]: Agree, in scope and would be great to have that merged. That is what the mention of Bluestein algorithm above is about.

File: doc/ROADMAP.rst.txt
Comment (rgommers) [inline doc/ROADMAP.rst.txt]: Yes, definitely. I'll see if I can get around to doing that sometime soon.

File: doc/ROADMAP.rst.txt
Comment (rgommers) [inline doc/ROADMAP.rst.txt]: I think gh-6495 needs to be merged. There's always more to do, but I'm not sure that anything needs to be mentioned here.

File: doc/ROADMAP.rst.txt
Comment (rgommers) [inline doc/ROADMAP.rst.txt]: I see no harm and some potential value in a ``'circ'`` mode.

File: doc/ROADMAP.rst.txt
Comment (rgommers) [inline doc/ROADMAP.rst.txt]: gh-5608 just added a new feature to ``signal.convolve`` right? IIRC there was a thing with normalization conventions or something similar. Can't find the details summarized anywhere though. I think it's a matter of someone sitting down and comparing those 3 functions. Probably doesn't need to be on this roadmap though, can open an issue for it.

File: doc/ROADMAP.rst.txt
Comment (rgommers) [inline doc/ROADMAP.rst.txt]: Indeed. Will change the text to just say "fix or deprecate".

**Comment by rgommers** [comment]
Updated this PR for review comments and changes in the last year. Will fix the merge conflicts later.

File: doc/ROADMAP.rst.txt
Comment (ev-br) [inline doc/ROADMAP.rst.txt]: typo: transpar**e**nt

File: doc/ROADMAP.rst.txt
Comment (ev-br) [inline doc/ROADMAP.rst.txt]: If that's mentioned, then also need to mention CJ's https://github.com/perimosocordiae/sparray

File: doc/ROADMAP.rst.txt
Comment (ev-br) [inline doc/ROADMAP.rst.txt]: Just wondering (possibly OT, but still). There are some extended precision helpers in special, used e.g. for  `log1p`, https://github.com/scipy/scipy/blob/master/scipy/special/c_misc/double2.h. Could they be useful for hyp1f1 et al?

**Comment by ev-br** [comment]
It might also make sense to open separate enhancement requests for simple well-defined easy-fix type tasks like moving `pade` and `logsumexp` to scipy.special. The motivation being that such issues are more visible for new contributors.

**Comment by rgommers** [comment]
> It might also make sense to open separate enhancement requests for simple well-defined easy-fix type tasks like moving pade and logsumexp to scipy.special. The motivation being that such issues are more visible for new contributors.

Agree, I can go through the whole thing after it's merged and see where that makes sense.

File: doc/ROADMAP.rst.txt
Comment (rgommers) [inline doc/ROADMAP.rst.txt]: good point, done

**Comment by rgommers** [comment]
Updated for Evgeni's comments, merged with master (wasn't rebase-able easily after 3 years anymore), and resolved some issues in the html doc build. Should be good to go now.

**Comment by rgommers** [comment]
I plan to merge this on Monday, unless something blocking shows up before then.

**Comment by ev-br** [comment]
> I plan to merge this on Monday, unless something blocking shows up before then.

+1

**Review by dlax — APPROVED** [review]

**Comment by rgommers** [comment]
Okay, in it goes. Thanks everyone. Now let's see if we can improve it over time....

**Comment by ev-br** [comment]
Yay!

**Pull request merged**
